### PR TITLE
#53621 Correct the schema for the `wp/v2/block-directory/search` endpoint

### PR DESCRIPTION
**Pull request by johnbillion**
Trac ticket: https://core.trac.wordpress.org/ticket/53621

### CODE REVIEW

**Comment by adamziel** [comment]
Here's a unit test that could be added to `tests/phpunit/tests/rest-api/rest-block-directory-controller.php`. I started off with some actual API data, but turns out there is a `get_mock_plugin()` that does essentially the same thing so I used that instead.

```php

	/**
	 * @ticket 53621
	 */
	public function test_get_items_response_conforms_to_schema() {
		wp_set_current_user( self::$admin_id );
		$plugin = $this->get_mock_plugin();
		$cb     = function () use ( $plugin ) {
			return (object) array(
				'info'    =>
					array(
						'page'    => 1,
						'pages'   => 1,
						'results' => 1,
					),
				'plugins' => array(
					$plugin,
				),
			);
		};
		add_filter( 'plugins_api', $cb );

		$request = new WP_REST_Request( 'GET', '/wp/v2/block-directory/search' );
		$request->set_query_params( array( 'term' => 'cache' ) );

		$expected = array(
			array(
				'name'                => 'sortabrilliant/guidepost',
				'title'               => 'Guidepost',
				'description'         => 'A guidepost gives you directions. It lets you know where you’re going. It gives you a preview of what’s to come.',
				'id'                  => 'guidepost',
				'rating'              => 4.3,
				'rating_count'        => 90,
				'active_installs'     => 100,
				'author_block_rating' => 0,
				'author_block_count'  => 1,
				'author'              => 'sorta brilliant',
				'icon'                => 'https://ps.w.org/guidepost/assets/icon-128x128.jpg?rev=2235512',
				'last_updated'        => gmdate( 'Y-m-d\TH:i:s', strtotime( $plugin['last_updated'] ) ),
				'humanized_updated'   => sprintf( '%s ago', human_time_diff( strtotime( $plugin['last_updated'] ) ) ),
				'_links'              => array(
					'wp:install-plugin' => array(
						array(
							'href' => 'http://example.org/index.php?rest_route=%2Fwp%2Fv2%2Fplugins&slug=guidepost',
						),
					),
					'curies'            => array(
						array(
							'name'      => 'wp',
							'href'      => 'https://api.w.org/{rel}',
							'templated' => true,
						),
					),
				),
			),
		);

		$result = rest_do_request( $request );
		remove_filter( 'plugins_api', $cb );

		$this->assertNotWPError( $result->as_error() );
		$this->assertSame( 200, $result->status );
		$this->assertSame( $expected, $result->get_data() );
	}
```

File: tests/phpunit/tests/rest-api/rest-block-directory-controller.php [hunk 1]
```diff
@@ -202,6 +202,70 @@ public function test_get_item_schema() {
...
+						),
+					),
+					'curies'            => array(
+						array(
+							'name'      => 'wp',
+							'href'      => 'https://api.w.org/{rel}',
+							'templated' => true,
+						),
+					),
+				),
```
Comment (TimothyBJacobs) [inline tests/phpunit/tests/rest-api/rest-block-directory-controller.php:257]: Let's unset the `_links` field so we don't need to worry about comparing correct URLs.

File: tests/phpunit/tests/rest-api/rest-block-directory-controller.php [hunk 2]
```diff
@@ -202,6 +202,70 @@ public function test_get_item_schema() {
...
+			),
+		);
+
+		$result = rest_do_request( $request );
+		remove_filter( 'plugins_api', $cb );
```
Comment (TimothyBJacobs) [inline tests/phpunit/tests/rest-api/rest-block-directory-controller.php:262]: We can remove this line, all filters get reset after each test.

**Comment by TimothyBJacobs** [comment]
This test is somewhat of a duplicate of our existing `test_prepare_item`. What we probably actually want here is to test that the REST API response validates against the JSON Schema for the endpoint.

We can do this by applying `rest_validate_value_from_schema` on the response data. We'll need to grab the schema for the endpoint which could be done by making an `OPTIONS` request like we do in `test_context_param` (preferred) or instantiating a `WP_REST_Block_Directory_Controller` instance and just calling `get_public_item_schema`.

**Comment by johnbillion** [comment]
I adjusted then test as per Timothy's recommendation. It now validates the result against the schema. It successfully fails with the old schema and passes with the corrected one.

**Review by adamziel — APPROVED** [review]

**Review by spacedmonkey — APPROVED** [review]

**Comment by spacedmonkey** [comment]
Looks good to me.

**Comment by johnbillion** [comment]
https://core.trac.wordpress.org/changeset/53315

**Pull request closed without merging**